yt_dlp/networking/common.py

```diff
@@ -0,0 +1,552 @@
+
+        client_certfile = self.ydl.params.get('client_certificate')
+        if client_certfile:
+            try:
+                context.load_cert_chain(
+                    client_certfile, keyfile=self.ydl.params.get('client_certificate_key'),
+                    password=self.ydl.params.get('client_certificate_password'))
+            except ssl.SSLError:
+                raise YoutubeDLError('Unable to load client certificate')
+
+        return context
+
+    def _make_sslcontext(self, verify, **kwargs):
+        """Generates a default HTTP/1.1 SSLContext with certs"""
+        context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
+        context.check_hostname = verify
+        context.verify_mode = ssl.CERT_REQUIRED if verify else ssl.CERT_NONE
+        # Some servers may reject requests if ALPN extension is not sent. See:
+        # https://github.com/python/cpython/issues/85140
+        # https://github.com/yt-dlp/yt-dlp/issues/3878
+        with contextlib.suppress(NotImplementedError):
+            context.set_alpn_protocols(['http/1.1'])
+        if verify:
+            ssl_load_certs(context, self.ydl.params)
+        return context
+
+    def _check_scheme(self, request: Request):
+        scheme = urllib.parse.urlparse(request.url).scheme.lower()
+        if scheme == 'file':  # no other handler should handle this request
+            raise RequestError('file:// scheme is explicitly disabled in yt-dlp for security reasons')
+
+        if self.SUPPORTED_SCHEMES is not None and scheme not in self.SUPPORTED_SCHEMES:
+            raise UnsupportedRequest(f'unsupported scheme: "{scheme}"')
+
+    def _check_proxies(self, request: Request):
+        if self.SUPPORTED_PROXY_SCHEMES is None:
+            return
+        for proxy_key, proxy_url in request.proxies.items():
+            if proxy_url is None:
+                continue
+            if proxy_key == 'no':
+                if Features.NO_PROXY not in self.SUPPORTED_FEATURES:
+                    raise UnsupportedRequest('\'no\' proxy is not supported')
+                continue
+            if proxy_key == 'all' and Features.ALL_PROXY not in self.SUPPORTED_FEATURES:
+                # XXX: If necessary, we could break up all_proxy here using SUPPORTED_SCHEMES
+                raise UnsupportedRequest('\'all\' proxy is not supported')
+
+            # Unlikely this handler will use this proxy, so ignore.
+            # This is to allow a case where a proxy may be set for a protocol
+            # for one handler in which such protocol (and proxy) is not supported by another handler.
+            if self.SUPPORTED_SCHEMES is not None and proxy_key not in self.SUPPORTED_SCHEMES + ['all']:
+                continue
+
+            scheme = urllib.parse.urlparse(proxy_url).scheme.lower()
+            if scheme not in self.SUPPORTED_PROXY_SCHEMES:
+                raise UnsupportedRequest(f'unsupported proxy type: "{scheme}"')
+
+    def prepare_request(self, request: Request):
```

Comment:
I'm not entirely confident in the design of this request flow.

There should to be a way to check if a handler can support a request or not (without it making the actual request if it does support).

Originally I had a `can_handle` function that returned True/False for this.
But this had issues:
- no easy way of getting the reason why the handler is not supported
- the request needs to be "modified" in a way before it can be supported.
    - we may find out while preparing the request that it is not supported
   
So then I went with a prepare request function then handle flow

The situation we have now:
- we have this weird public prepare_request function that modifies the given Request.. with no use of the modification outside the RH
- handle() function which calls prepare_request function internally to get the modified Request

One way I was thinking we could clean this up could be by doing:
- make prepare_request internal only
- Have a `can_handle` function that runs a given Request through prepare_request, but doesn't return a Request
    - when it can't handle, do we want to raise the UnsupportedRequest error with a reason or return a bool?

but I am open to other suggestions too